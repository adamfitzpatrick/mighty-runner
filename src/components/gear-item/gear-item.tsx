import * as React from 'react'

import * as styles from './gear-item.scss'
import * as Models from '@models'
import { EditItemRenderProp } from '@components/edit-item'
import { Effects } from '@containers/effects'

interface Props {
  item: Models.GearItem
  edit: (item: Models.GearItem) => void
}

export default function GearItem ({ item, edit }: Props) {

  return (
    <div>
      <div>
        <h3>{ item.name }</h3>
        <div>Description: { item.description }</div>
        <div>Cost: { item.cost }</div>
        <div>Availability: { item.availability }</div>
        <div>Effects:</div>
        <Effects effectIds={item.effects} />
        <button onClick={ () => edit(item) }>Edit</button>
      </div>
    </div>
  )
}

export const GearItemEditRender: EditItemRenderProp<Models.GearItem> = (item, changeHandler) => {
  const onChangeCreator = (property: keyof Models.GearItem, num?: boolean) => {
    return (event: React.ChangeEvent<HTMLInputElement>) => {
      const value = num ? parseFloat(event.target.value) || 0 : event.target.value
      const updated = {
        ...item,
        [property]: value
      }
      changeHandler(updated)
    }
  }

  return (
    <div>
      <label>
        <span>Name</span>
        <input
          value={item.name}
          onChange={onChangeCreator('name')}
        />
      </label>
      <label>
        <span>Description</span>
        <input
          value={item.description}
          onChange={onChangeCreator('description')}
        />
      </label>
      <label>
        <span>Cost</span>
        <input
          value={item.cost}
          onChange={onChangeCreator('cost', true)}
        />
      </label>
      <label>
        <span>Availability</span>
        <input
          value={item.availability}
          onChange={onChangeCreator('availability')}
        />
      </label>
    </div>
  )
}
