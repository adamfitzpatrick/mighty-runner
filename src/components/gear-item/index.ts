export { default } from './gear-item'
export * from './gear-item'
