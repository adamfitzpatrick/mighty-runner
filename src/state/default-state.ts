import { Character, PersonalData, CharacterIdentifier, Attributes, Effect, GearItem, Pic } from '@models'

export interface AppState {
  characters: Character[] | null
  activeCharacter: CharacterIdentifier | null
  pic: Pic | null
  personalData: PersonalData | null
  attributes: Attributes | null
  gear: GearItem[] | null
  effects: Effect[] | null
  apiError: boolean
}

export const defaultState: AppState = {
  characters: null,
  pic: null,
  activeCharacter: null,
  personalData: null,
  attributes: null,
  gear: null,
  effects: null,
  apiError: false
}

export default defaultState
