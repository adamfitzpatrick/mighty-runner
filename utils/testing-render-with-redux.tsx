import * as React from 'react'
import {
  createStore,
  combineReducers,
  AnyAction,
  Store,
  Dispatch,
  applyMiddleware
} from 'redux'
import { Provider } from 'react-redux'
import { render, RenderResult } from '@testing-library/react'
import * as Models from '@models'
import { AppState } from '../src/state/default-state'

export type RenderWithRedux =
  RenderResult &
  {
    store: Store<AppState, AnyAction>,
    dispatchSpy: jest.Mock<Dispatch<AnyAction>>
  }

const reducer = combineReducers({
  characters: (state: Models.Character[] | null = null) => state,
  activeCharacter: (state: Models.CharacterIdentifier | null = null) => state,
  personalData: (state: Models.PersonalData | null = null) => state,
  attributes: (state: Models.Attributes | null = null) => state,
  gear: (state: Models.GearItem[] | null = null) => state,
  effects: (state: Models.Effect[] | null = null) => state
})

const getSpyMiddleware = (dispatchSpy: jest.Mock<Dispatch<AnyAction>>) => () => {
  return (next: Dispatch<AnyAction>) => (action: AnyAction) => {
    dispatchSpy(action)
    next(action)
  }
}

export function renderWithRedux (
  ui: JSX.Element,
  initialState: AppState
): RenderWithRedux {
  const dispatchSpy: jest.Mock<Dispatch<AnyAction>> = jest.fn()
  const spyMiddleware = getSpyMiddleware(dispatchSpy)
  const store = createStore(reducer, initialState, applyMiddleware(spyMiddleware))

  return {
    ...render(<Provider store={store}>{ ui }</Provider>),
    store,
    dispatchSpy
  }
}
