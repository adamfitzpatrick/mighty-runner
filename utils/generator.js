const fs = require('fs')
const path = require('path')
const yargs = require('yargs')

if (!yargs.argv.container && !yargs.argv.component) {
  throw new Error('no component or container name provided')
}

const container = yargs.argv.container && yargs.argv.container.replace(/^[a-z]/, a => a.toUpperCase())
const component = yargs.argv.component && yargs.argv.component.replace(/^[a-z]/, a => a.toUpperCase())

let thing = yargs.argv.component
if (container) {
  thing = container
}
const lower = thing.replace(/^[A-Z]/, a => a.toLowerCase()).replace(/[A-Z]/g, a => `-${a.toLowerCase()}`)
thing = thing.replace(/^[a-z]/, a => a.toUpperCase()).replace(/-([A-Za-z0-9])/g, a => a.slice(1).toUpperCase())
scssClass = thing.replace(/^[A-Z]/, a => a.toLowerCase())

const thingTs = `import * as React from 'react'

import * as styles from './${lower}.scss'

interface Props {}

export default function ${thing} (props: Props) {
  return <div>${thing}</div>
}
`

const thingTestTs = `import * as React from 'react'
import { cleanup, render } from '@testing-library/react'
import { renderWithRedux, RenderWithRedux } from '../../../utils/testing-render-with-redux'
import ${thing} from '.'

describe('${thing} ${ container ? 'container' : 'component'}', () => {
  test('should do nothing', () => {
    expect(true).toEqual(true)
  })
})
`

const indexTs = `export { default } from './${lower}'
`

const scss = `.${scssClass} {
  display: block;
}
`

const scssD = `export const ${scssClass}: string
`

let dest = path.resolve(process.cwd(), 'src', 'components', lower)
if (container) {
  dest = path.resolve(process.cwd(), 'src', 'containers', lower)
}

fs.mkdirSync(dest)
fs.writeFileSync(path.resolve(dest, `${lower}.tsx`), thingTs)
fs.writeFileSync(path.resolve(dest, `${lower}.test.tsx`), thingTestTs)
fs.writeFileSync(path.resolve(dest, 'index.ts'), indexTs)
fs.writeFileSync(path.resolve(dest, `${lower}.scss`), scss)
fs.writeFileSync(path.resolve(dest, `${lower}.scss.d.ts`), scssD)
