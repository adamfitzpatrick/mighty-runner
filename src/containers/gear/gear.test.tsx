import * as React from 'react'
import { renderWithRedux, RenderWithRedux } from '../../../utils/testing-render-with-redux'
import Gear from '.'
import { AppState } from '@state/default-state'
import * as Models from '@models'
import { cleanup, fireEvent } from '@testing-library/react'
import { gearItem } from '@components/gear-item/gear-item.scss'

describe('Gear container', () => {
  let appState: AppState
  let gear: Models.GearItem[]
  let effects: Models.Effect[]
  let sut: RenderWithRedux

  beforeEach(() => {
    gear = [{
      id: '1',
      name: 'n',
      shortName: 's',
      description: 'd',
      asEffectTarget: [ 'a', 'b' ],
      effects: [ '1' ],
      cost: 1,
      availability: '6R',
      value: { rating: 3 }
    }]
    effects = [{
      id: '1',
      name: 'n',
      description: 'd',
      target: [ 'c', 'd' ],
      value: 1,
      active: true
    }]
    appState = { gear, effects } as AppState
    sut = renderWithRedux(<Gear />, appState)
  })

  test('should render correctly', () => {
    expect(sut.container.innerHTML).toMatchSnapshot()
  })

  test('should not throw an error if application state is unavailable', () => {
    cleanup()
    delete appState.gear
    sut = renderWithRedux(<Gear />, appState)
    expect(sut.container.innerHTML).toMatchSnapshot()
  })

  describe('editing gear', () => {
    beforeEach(() => {
      const edit = sut.getByTestId('Edit.button.component')
      fireEvent.click(edit)
    })

    test('should add editing component when the edit button is clicked', () => {
      expect(sut.getByTestId('n.edit-item.component')).toBeDefined()
    })

    test('should store changes locally until done button is clicked', () => {
      const name = sut.getAllByTestId('Name.input.component')[0]
      fireEvent.change(name, { target: { value: 'changed' } })
    })

    test('should save changes when the done handler fires', () => {
      const done = sut.getByTestId('Done.button.component')
      fireEvent.click(done)
      expect(() => sut.getByTestId('n.edit-item.component')).toThrow()
      const dispatches = sut.dispatchSpy.mock.calls.map(call => call[0])
      expect(dispatches[0].type).toEqual('UPDATE_GEAR')
      expect(dispatches[1].type).toEqual('SAVE_CHARACTER')
    })
  })

  describe('removing gear', () => {
    test('should remove the specified gearItem', () => {
      const remove = sut.getAllByTestId('Delete.button.component')[0]
      fireEvent.click(remove)
      expect(sut.dispatchSpy).toHaveBeenCalledWith({
        payload: gear[0],
        type: 'REMOVE_GEAR'
      })
      expect(sut.dispatchSpy).toHaveBeenCalledWith({
        type: 'SAVE_CHARACTER'
      })
    })
  })

  describe('adding gear', () => {
    beforeEach(() => {
      const add = sut.getByTestId('Add.button.component')
      fireEvent.click(add)
    })

    test('should add editing component when the add button is clicked', () => {
      expect(sut.getByTestId('.edit-item.component')).toBeDefined()
    })

    test('should store additions locally until done button is clicked', () => {
      const name = sut.getByTestId('Name.input.component')
      fireEvent.change(name, { target: { value: 'changed' } })
      expect(sut.getByTestId('changed.edit-item.component')).toBeDefined()
    })

    test('should not save additions when the cancel button is clicked', () => {
      const cancel = sut.getByTestId('Cancel.button.component')
      fireEvent.click(cancel)
      expect(sut.dispatchSpy).not.toHaveBeenCalled()
    })

    test('should persist additions when the done button is clicked', () => {
      const name = sut.getByTestId('Name.input.component')
      fireEvent.change(name, { target: { value: 'item' } })
      const done = sut.getByTestId('Done.button.component')
      fireEvent.click(done)
      expect(sut.dispatchSpy).toHaveBeenCalledWith({
        payload: expect.objectContaining({
          name: 'item'
        }),
        type: 'ADD_GEAR'
      })
      expect(sut.dispatchSpy).toHaveBeenCalledWith({ type: 'SAVE_CHARACTER' })
    })
  })
})
