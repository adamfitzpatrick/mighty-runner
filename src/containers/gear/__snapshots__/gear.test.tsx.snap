// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gear container should not throw an error if application state is unavailable 1`] = `""`;

exports[`Gear container should render correctly 1`] = `"<div><h2>Gear</h2><div data-testid=\\"n.gear-item.component\\"><h3>n</h3><div>Description: d</div><div>Cost: 1</div><div>Rating: 3</div><div>Availability: 6R</div><div>Effects:</div><div data-testid=\\"non-editable-effect.component\\"><span>Name: n</span><label><span>Active</span><input data-testid=\\"Active.input.component\\" type=\\"checkbox\\" value=\\"\\" checked=\\"\\"></label></div><button data-testid=\\"Edit.button.component\\">Edit</button><button data-testid=\\"Delete.button.component\\">Delete</button></div><button data-testid=\\"Add.button.component\\">Add</button></div>"`;
